src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -0,0 +1,447 @@
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.NewBlock
+import io.iohk.ethereum.network.p2p.messages.PV62._
+import io.iohk.ethereum.network.p2p.messages.PV63.{GetNodeData, NodeData}
+import io.iohk.ethereum.network.p2p.{Message, MessageSerializable}
+import io.iohk.ethereum.utils.Config.SyncConfig
+import mouse.all._
+
+import scala.collection.immutable.Queue
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+
+class BlockFetcher(
+    val etcPeerManager: ActorRef,
+    val peerEventBus: ActorRef,
+    val syncConfig: SyncConfig,
+    implicit val scheduler: Scheduler
+) extends Actor
+    with ActorLogging
+    with PeerListSupport
+    with BlacklistSupport {
+
+  import BlockFetcher._
+
+  val bufferSize = 10000
+
+  val printSchedule: Cancellable = scheduler.schedule(1.second, 1.second, self, PrintStatus)
+  peerEventBus ! Subscribe(MessageClassifier(Set(NewBlock.code, NewBlockHashes.code), PeerSelector.AllPeers))
+
+  override def receive: Receive = idle()
+
+  override def postStop(): Unit = {
+    super.postStop()
+    printSchedule.cancel()
+    peerEventBus ! Unsubscribe()
+  }
+
+  private def idle(): Receive =
+    handleCommonMessages(None) orElse {
+      case Start(importer, blockNr) => FetcherState.initial(importer, blockNr) |> fetchBlocks
+    }
+
+  def handleCommonMessages(state: Option[FetcherState]): Receive =
+    handlePeerListMessages orElse handleBlacklistMessages orElse {
+      case PrintStatus =>
+        log.debug("BlockFetcher status {}", state.map(_.status))
+    }
+
+  private def started(state: FetcherState): Receive =
+    handleCommonMessages(Some(state)) orElse
+      handleCommands(state) orElse
+      handleNewBlockMessages(state) orElse
+      handleHeadersMessages(state) orElse
+      handleBodiesMessages(state) orElse
+      handleStateNodeMessages(state) orElse {
+      case msg => log.debug("Unhandled msg {} from {}", msg, sender())
+    }
+
+  private def handleCommands(state: FetcherState): Receive = {
+    case PickBlocks(amount) =>
+      val (blocks, newState) = state.pickBlocks(amount)
+      if (blocks.nonEmpty) {
+        sender() ! PickedBlocks(blocks.toList)
+      }
+      val topMsg = if (state.isOnTop) OnTop else NotOnTop
+      newState.importer ! topMsg
+      fetchBlocks(newState)
+    case InvalidateBlocksFrom(blockNr, reason) =>
+      val (blockProvider, newState) = state.invalidateBlocksFrom(blockNr)
+      blacklist(blockProvider.id, syncConfig.blacklistDuration, reason)
+      fetchBlocks(newState)
+    case FetchBlocks => fetchBlocks(state)
+  }
+
+  private def handleHeadersMessages(state: FetcherState): Receive = {
+    case ResponseReceived(peer, BlockHeaders(headers), _) if state.isHeadersFetcher(sender()) =>
+      log.debug("Fetched headers for blocks {} by {}", headers.map(_.number).mkString(","), sender())
+
+      val newState = state.validatedHeaders(headers) match {
```

Comment:
I do not see anywhere code to resolve branches - in old sync it was done by method `Ledger.resolveBranches`. 
Main idea is that you can have current best imported chain `(blocknumber, blockHash, parentHash)`:
`Genesis -> (1 , a, genesis)-> (2, b, a) -> (3,  c, b) `
but canonical chain now looks like:
`Genesis -> (1 , a, genesis)-> (2, d, a) -> (3,  e, d) `
so when you now receive header `(4, f, e)` ,  it will now fail vaildation , as we do not have its parent in blockchain. Worse, we will receive this header from all peers (as it is now canonical) and we wont be able move forward. 
As you look at old regular sync in that situation, we were asking peer for `n` headers  from current best to best - n , and then we were looking for common parent to determine best chain.